lib/IRGen/GenClangDecl.cpp

```diff
@@ -41,7 +41,6 @@ void IRGenModule::emitClangDecl(const clang::Decl *decl) {
   if (!valueDecl || valueDecl->isExternallyVisible()) {
     ClangCodeGen->HandleTopLevelDecl(
                           clang::DeclGroupRef(const_cast<clang::Decl*>(decl)));
-    return;
   }
```

Comment:
This looks correct, but can we preserve the fast path where the declaration does not have a definition?

Reply:
I assume you mean the fast path where the declaration isn't a `ValueDecl` (which I've done)?

There wasn't a fast path before for the case where the declaration doesn't have a definition (and I believe this isn't possible to check in the general case of a `ValueDecl`?).

Reply:
Well, the fast-path was buggy, we've established that. :)  I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.  I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

Reply:
> Well, the fast-path was buggy, we've established that. :) I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Ah -- that makes sense.

> Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.

Oooooh, right! I've verified this case was not being handled correctly either. See the extra test case I've added -- `calledTransitivelyFromVarInit()` was not being emitted before I made the code changes in my latest commit.

> I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

A `VarDecl` doesn't have a `Stmt *` though...? But I see what you mean, and I've introduced a helper function that returns a redeclaration of the decl containing "executable code" (body for a function, initializer for a variable), if one exists. This decl contains the body (in the case of a function) or initializer (in the case of a variable), so it's the right thing to traverse to find DeclRefs of things we should emit.

Not super-pleased with the naming of the helper function BTW (`getDeclWithExecutableCode()`) -- let me know if you have a better suggestion.

Reply:
A quirk of Clang's AST is that `Expr` is a subclass of `Stmt`, so `VarDecl`s *do* actually have a `Stmt*`, but this is a fine approach, too.

Reply:
Oh right, I had forgotten about that -- makes sense, given that an expression is a valid statement in C++.